File: tools/verify-scenes.js
#!/usr/bin/env node
/**
 * Static smoke check for js/game-scenes.js: every choice.next points at a real scene
 * (or ending_computed), crisis rows use roll + null next, RESOLVE_PATH_LEAD keys match paths.
 *
 * Run from repo root (Final/):  node tools/verify-scenes.js
 */

const fs = require("fs");
const path = require("path");
const vm = require("vm");

const root = path.join(__dirname, "..");
const scenesPath = path.join(root, "js", "game-scenes.js");
const configPath = path.join(root, "js", "game-config.js");

function loadScenes() {
  const code = fs.readFileSync(scenesPath, "utf8");
  const ctx = {};
  vm.createContext(ctx);
  vm.runInContext(code.replace(/^\s*const scenes =/, "var scenes ="), ctx);
  if (!ctx.scenes || typeof ctx.scenes !== "object") {
    throw new Error("Failed to load scenes object from game-scenes.js");
  }
  return ctx.scenes;
}

function loadResolveLeadKeys() {
  const config = fs.readFileSync(configPath, "utf8");
  const start = config.indexOf("const RESOLVE_PATH_LEAD");
  if (start < 0) throw new Error("RESOLVE_PATH_LEAD not found in game-config.js");
  const rest = config.slice(start);
  const endMarker = rest.indexOf("\n    };");
  const block = endMarker >= 0 ? rest.slice(0, endMarker) : rest;
  const required = ["west", "slav", "statist", "med"];
  const missing = required.filter((k) => !new RegExp(`\\n\\s*${k}\\s*:`).test(block));
  if (missing.length) {
    throw new Error(`RESOLVE_PATH_LEAD missing keys: ${missing.join(", ")}`);
  }
  return required;
}

function loadSchemeSceneKeys() {
  const config = fs.readFileSync(configPath, "utf8");
  const start = config.indexOf("const SCENE_COLOR_SCHEME");
  if (start < 0) return [];
  const rest = config.slice(start);
  const m = rest.match(/const SCENE_COLOR_SCHEME = \{([\s\S]*?)\n    \};/);
  if (!m) return [];
  const body = m[1];
  const keys = [];
  const re = /\n\s*([a-z0-9_]+)\s*:/g;
  let x;
  while ((x = re.exec(body))) keys.push(x[1]);
  return keys;
}

function main() {
  const errors = [];
  const warnings = [];

  const scenes = loadScenes();
  const sceneKeys = new Set(Object.keys(scenes));

  try {
    loadResolveLeadKeys();
  } catch (e) {
    errors.push(e.message);
  }

  for (const sk of loadSchemeSceneKeys()) {
    if (!sceneKeys.has(sk)) {
      errors.push(`SCENE_COLOR_SCHEME references unknown scene "${sk}"`);
    }
  }

  for (const [id, scene] of Object.entries(scenes)) {
    if (!scene || typeof scene !== "object") continue;
    const choices = scene.choices;
    if (!Array.isArray(choices)) {
      if (!scene.computed) warnings.push(`Scene "${id}" has no choices array`);
      continue;
    }
    choices.forEach((ch, i) => {
      if (!ch || typeof ch !== "object") return;
      const hasRoll = ch.roll === true;
      const n = ch.next;
      if (hasRoll) {
        if (n != null) errors.push(`Scene "${id}" choice[${i}]: roll choice must use next: null`);
        if (!String(id).startsWith("crisis_")) {
          warnings.push(`Scene "${id}" choice[${i}]: roll:true outside crisis_* (unexpected)`);
        }
        return;
      }
      if (n == null || n === "") {
        errors.push(`Scene "${id}" choice[${i}]: missing next (not a roll choice)`);
        return;
      }
      if (typeof n !== "string") {
        errors.push(`Scene "${id}" choice[${i}]: next must be a string`);
        return;
      }
      if (n === "ending_computed") return;
      if (!sceneKeys.has(n)) {
        errors.push(`Scene "${id}" choice[${i}]: unknown next "${n}"`);
      }
    });
  }

  if (!sceneKeys.has("intro")) errors.push('Missing scene "intro"');
  if (!sceneKeys.has("ending_computed")) errors.push('Missing scene "ending_computed"');

  const referenced = new Set();
  referenced.add("intro");
  for (const scene of Object.values(scenes)) {
    if (!scene || !Array.isArray(scene.choices)) continue;
    for (const ch of scene.choices) {
      if (ch && typeof ch.next === "string" && ch.next !== "ending_computed") referenced.add(ch.next);
    }
  }
  const eventJump = new Set([
    "event_censor",
    "event_flood_echo",
    "event_salon",
    "event_rural_gentry",
    "event_zemstvo_clash"
  ]);
  for (const id of sceneKeys) {
    if (!referenced.has(id) && id !== "ending_computed" && !eventJump.has(id)) {
      warnings.push(`Scene "${id}" is never targeted by any choice.next (orphan / dev only?)`);
    }
  }

  if (warnings.length) {
    console.warn("Warnings:\n", warnings.map((w) => `  - ${w}`).join("\n"));
  }
  if (errors.length) {
    console.error("Errors:\n", errors.map((e) => `  - ${e}`).join("\n"));
    process.exit(1);
  }

  console.log(
    `verify-scenes: OK; ${sceneKeys.size} scenes, ${referenced.size} referenced from choices, RESOLVE_PATH_LEAD keys present.`
  );
}

main();

File: tools/verify-scenes.py
#!/usr/bin/env python3
"""
Static smoke check for js/game-scenes.js (same intent as verify-scenes.js).
Run from Final/:  python tools/verify-scenes.py
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SCENES = ROOT / "js" / "game-scenes.js"
CONFIG = ROOT / "js" / "game-config.js"

# Reached via winter roll (pickCrisisEvent), not a static choice.next string.
EVENT_JUMP_TARGETS = {
    "event_censor",
    "event_flood_echo",
    "event_salon",
    "event_rural_gentry",
    "event_zemstvo_clash",
}


def main() -> int:
    errors: list[str] = []
    warnings: list[str] = []

    text = SCENES.read_text(encoding="utf-8")
    scene_keys = set(re.findall(r"^\s{6}([a-z0-9_]+):\s*\{", text, re.MULTILINE))

    # Choice lines are single-line objects with "text:" (avoids primaryRead bodies).
    for i, line in enumerate(text.splitlines(), 1):
        stripped = line.strip()
        if not stripped.startswith("{ text:"):
            continue
        if "roll: true" in line or ", roll: true" in line:
            if "next: null" not in line:
                errors.append(f"line {i}: roll choice must have next: null")
            continue
        m = re.search(r'next:\s*"([a-z0-9_]+)"', line)
        if not m:
            errors.append(f"line {i}: choice missing next (expected next: \"…\")")
            continue
        target = m.group(1)
        if target != "ending_computed" and target not in scene_keys:
            errors.append(f"line {i}: unknown next \"{target}\"")

    cfg = CONFIG.read_text(encoding="utf-8")
    start = cfg.find("const RESOLVE_PATH_LEAD")
    if start < 0:
        errors.append("RESOLVE_PATH_LEAD not found in game-config.js")
    else:
        block = cfg[start : start + 8000]
        for k in ("west", "slav", "statist", "med"):
            if not re.search(rf"\n\s*{k}\s*:", block):
                errors.append(f"RESOLVE_PATH_LEAD missing key {k!r}")

    sch = re.search(
        r"const SCENE_COLOR_SCHEME = \{(.*?)\n    \};", cfg, re.DOTALL
    )
    if sch:
        for sk in re.findall(r"\n\s*([a-z0-9_]+)\s*:", sch.group(1)):
            if sk not in scene_keys:
                errors.append(f'SCENE_COLOR_SCHEME references unknown scene "{sk}"')

    if "intro" not in scene_keys:
        errors.append('Missing scene "intro"')
    if "ending_computed" not in scene_keys:
        errors.append('Missing scene "ending_computed"')

    referenced = {"intro", "session_format"}
    for m in re.finditer(r'next:\s*"([a-z0-9_]+)"', text):
        t = m.group(1)
        if t != "ending_computed":
            referenced.add(t)

    for sid in sorted(scene_keys):
        if (
            sid != "ending_computed"
            and sid not in referenced
            and sid not in EVENT_JUMP_TARGETS
        ):
            warnings.append(
                f'Scene "{sid}" is never targeted by any choice.next (orphan / dev only?)'
            )

    if warnings:
        print("Warnings:\n" + "\n".join(f"  - {w}" for w in warnings), file=sys.stderr)
    if errors:
        print("Errors:\n" + "\n".join(f"  - {e}" for e in errors), file=sys.stderr)
        return 1

    print(
        "verify-scenes: OK - "
        f"{len(scene_keys)} scenes, {len(referenced)} referenced from choices, "
        "RESOLVE_PATH_LEAD keys present."
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
